tests/quantization/bnb/test_replace_linear.py

```diff
@@ -0,0 +1,220 @@
+            self.submodel_nested_2 = NestedModel()
+
+    def count_modules_by_names(model: nn.Module, target_names: Optional[List[str]]) -> int:
+        """
+        Count all submodules within a PyTorch model that have names specified in the target_names list.
+        """
+        if not target_names:
+            return 0
+
+        return sum(1 for name, _ in model.named_modules() if name.split(".")[-1] in target_names)
+
+
+@dataclass
+class TestConfig:
+    bnb_cfg: BitsAndBytesConfig
+    replacement_linear_class: Type
+    shorthand: str
+    modules_to_not_convert: Optional[List[str]] = None
+    llm_int8_skip_modules: Optional[List[str]] = None
+
+
+@require_bitsandbytes
+class TestReplaceWithBnbLinear(unittest.TestCase):
+    def setUp(self):
+        self.mock_log_warning_patch = patch("logging.Logger.warning")
+        self.mock_log_warning = self.mock_log_warning_patch.start()
+        self.warning_checked = False
+
+    def tearDown(self):
+        self.mock_log_warning_patch.stop()
+
+        if not self.warning_checked and self.mock_log_warning.call_args_list:
+            unexpected_warnings = [call_args[0][0] for call_args in self.mock_log_warning.call_args_list]
+            formatted_warnings = "\n   ".join(["Unexpected warning(s) logged:"] + unexpected_warnings)
+
+            self.fail(formatted_warnings)
+
+    def check_warning_logged(self, warning_message):
+        self.warning_checked = True
+
+        found = any(warning_message in call_args[0][0] for call_args in self.mock_log_warning.call_args_list)
+        self.assertTrue(found, f"Expected warning message '{warning_message}' was not found")
+
+    def verify_replaced_module_count(self, model, replaced_class, expected_count):
+        count = sum(1 for module in model.modules() if isinstance(module, replaced_class))
+        self.assertEqual(count, expected_count, f"{expected_count} modules should be replaced")
+
+    def common_test_logic(
+        self,
+        test_cfg: TestConfig,
+        dummy_model_class,
+        expected_replaced_count,
+        expected_warning_message=None,
+    ):
+        model = dummy_model_class()
+        replaced_model = replace_with_bnb_linear(
+            model,
+            quantization_config=test_cfg.bnb_cfg,
+            modules_to_not_convert=test_cfg.modules_to_not_convert,
+        )
+
+        self.verify_replaced_module_count(
+            replaced_model,
+            test_cfg.replacement_linear_class,
+            expected_replaced_count,
+        )
+
+        for name, module in replaced_model.named_modules():
+            if isinstance(module, test_cfg.replacement_linear_class):
+                self.assertIsInstance(module, test_cfg.replacement_linear_class)
+
+            elif "lm_head" in name:
+                self.assertEqual(type(module), Linear)
+
+        if expected_warning_message:
+            self.check_warning_logged(expected_warning_message)
+
+    @staticmethod
+    def add_test_case(
```

Comment:
Hi @younesbelkada,

Thanks for your review! I concur, the tests might seem complex initially, but they are structured into small, clear segments for better understanding.

Here are my thoughts:
- To cover all code paths in `replace_linear`, parametrization was essential. Since this is the key integration point between bnb and transformers, a comprehensive unit test for all scenarios seemed necessary.
- `unittest` doesn't allow for easy parameterization and "complicated" logic is in the end a workaround for that. With `pytest` these tests could look much cleaner.
- As we expand the quantization config settings, incorporating new behaviors becomes straightforward with the current setup. By simply adding a TestConfig to `quantization_test_configs()`, we can seamlessly integrate and validate new code. I’ve now added an explanatory doc-string to this function.
- My goal was to make it super simple to add new replacement types, accepting the complexity in testing as a deliberate trade-off, especially under the constraint of using `unittest`.
- Rewriting the tests would require a non-trivial effort, which I’d prefer to allocate differently for now. I anticipate a future refactor, possibly post-modeling_utils merge or when introducing new replacement types.
- I believe we should proceed with the existing tests, making iterative improvements as necessary in future versions.

What are your thoughts?

Reply:
Makes sense - However I think that you can just implement 3 simple tests (one without any linear / conv1d, one with linear, and one with conv1d and it would cover most of the usecases right?) @amyeroberts @ArthurZucker can confirm but I think in transformers we prefer to have simple tests for better readability

Reply:
yep I don't think we need an extra testing file here. it' s good to test averything but better if we can see the usage of this new feature pretty easily. 
-> test that we can replace with a real word use-case that should be sufficient to check edge cases!

Reply:
@ArthurZucker Thanks for the review! Sure, I'll integrate our feedback and come forward with a new revision. Just to understand, "extra testing file here": What do you mean by that?

Reply:
Do you mean that the whole test_replace_linear.py is not needed in your opinion?

The reason I created a separate file is that replace_linear was just implicitly tested before, with those tests running for 1h+.

I needed sth to quickly iterate, that runs in just a few seconds and makes it quick to validate the integration point with BNB or other replacement Linear layers.

Of course, I'll adapt the tests to your specifications, was just asking to clarify my understanding of your input.

Reply:
Yeah I think you can just add a test in the main testing file and select it with `pytest ... -k "test_replace_xxxx"` or whatever 🤗